src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Common/SourceTextHelper.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Common;
+internal static class SourceTextHelper
+{
+    public static bool IsLinePositionValid(this SourceText sourceText, LinePosition linePosition, ILogger? logger)
```

Comment:
We already have code that does most of this. eg, https://github.com/dotnet/razor/blob/main/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Extensions/LinePositionExtensions.cs#L15 which calls https://github.com/dotnet/razor/blob/main/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Extensions/SourceTextExtensions.cs#L206

Importantly, the latter method there is the (hopefully!) single source of truth for validity checks, and allows for a quirk of LSP that allows for the line to be 1 greater than the number of lines.

Having said that, I think its still good to have an `IsValid` method like [this one](https://github.com/dotnet/razor/blob/main/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Extensions/PositionExtensions.cs#L57) we have in `PositionExtensions`. Perhaps in `LinePositionExtensions`, or `SourceTextExtensions`?

Reply:
Thanks David, looked for it but somehow still missed it. 

One thing I'm noticing is that "character" is not getting checked by either Roslyn in SourceText or by us. Seems like we should check for that and not allow it be beyond the end of the line, no? Otherwise we don't detect correctly if there is a sync issue (as there appears to be in this case) and when the caller (editor) is asking for data at a bad position (beyond the end of the line) we should detect it, which we currently don't do.

Reply:
Good catch

Reply:
Combined the code and removed quite a bit of duplicate extension methods in Microsoft.VisualStudio.Razor.LanguageServerClient

Checking the character position in the line exposed several issues in our unit tests.